ast/edit/match.go

```diff
@@ -0,0 +1,436 @@
+		return matchArrayExpression(n, node.(*ast.ArrayExpression))
+	case *ast.Identifier:
+		return matchIdentifier(n, node.(*ast.Identifier))
+	case *ast.PipeLiteral:
+		return matchPipeLiteral(n, node.(*ast.PipeLiteral))
+	case *ast.StringLiteral:
+		return matchStringLiteral(n, node.(*ast.StringLiteral))
+	case *ast.BooleanLiteral:
+		return matchBooleanLiteral(n, node.(*ast.BooleanLiteral))
+	case *ast.FloatLiteral:
+		return matchFloatLiteral(n, node.(*ast.FloatLiteral))
+	case *ast.IntegerLiteral:
+		return matchIntegerLiteral(n, node.(*ast.IntegerLiteral))
+	case *ast.UnsignedIntegerLiteral:
+		return matchUnsignedIntegerLiteral(n, node.(*ast.UnsignedIntegerLiteral))
+	case *ast.RegexpLiteral:
+		return matchRegexpLiteral(n, node.(*ast.RegexpLiteral))
+	case *ast.DurationLiteral:
+		return matchDurationLiteral(n, node.(*ast.DurationLiteral))
+	case *ast.DateTimeLiteral:
+		return matchDateTimeLiteral(n, node.(*ast.DateTimeLiteral))
+	case *ast.FunctionExpression:
+		return matchFunctionExpression(n, node.(*ast.FunctionExpression))
+	case *ast.Property:
+		return matchProperty(n, node.(*ast.Property))
+	default:
+		// If we were able not to find the type, than this switch is wrong
+		panic(fmt.Errorf("unknown type %q", n.Type()))
+	}
+}
+
+func matchString(s, os string) bool {
+	if len(s) > 0 && s != os {
+		return false
+	}
+	return true
+}
+
+// negative operators are ignored (invalid value for enumeration)
+func matchOperator(o, oo ast.OperatorKind) bool {
+	if o >= 0 && o != oo {
+		return false
+	}
+	return true
+}
+
+func matchLogicalOperator(o, oo ast.LogicalOperatorKind) bool {
+	if o >= 0 && o != oo {
+		return false
+	}
+	return true
+}
+
+func matchPackage(p *ast.Package, op *ast.Package) bool {
+	if !matchString(p.Path, op.Path) ||
+		!matchString(p.Package, op.Package) {
+		return false
+	}
+	if len(p.Files) > len(op.Files) {
+		return false
+	}
+	if len(p.Files) == 0 {
+		return true
+	}
+	var matched bool
+	for i := 0; i < len(op.Files) && !matched; i++ {
+		if len(p.Files) > len(op.Files)-i-1 {
+			// stop searching, the sub-sequence is longer than the rest
+			break
+		}
+		for ii := 0; i < len(p.Files) && !matched; ii++ {
+			if !match(p.Files[ii], op.Files[i+ii]) {
+				// match not found, continue searching
+				break
+			}
+			if ii == len(p.Files)-1 {
+				// match found
+				matched = true
+			}
+		}
+	}
+	return matched
+}
+
+func matchFile(f *ast.File, of *ast.File) bool {
+	if !matchString(f.Name, of.Name) {
+		return false
+	}
+	if !match(f.Package, of.Package) {
+		return false
+	}
+	if len(f.Imports) > len(of.Imports) {
+		return false
+	}
+	if len(f.Imports) == 0 {
+		return true
+	}
+	var matched bool
+	for i := 0; i < len(of.Imports) && !matched; i++ {
+		if len(f.Imports) > len(of.Imports)-i-1 {
+			break
+		}
+		for ii := 0; i < len(f.Imports) && !matched; ii++ {
+			if !match(f.Imports[ii], of.Imports[i+ii]) {
+				break
+			}
+			if ii == len(f.Imports)-1 {
+				matched = true
+			}
+		}
+	}
+	return matched
```

Comment:
Do you not recurse into the `Body` statements here?

It seems like the matcher will most often be used at the level of the statements in `File.Body` and below... Do we need to be able to match the structure of a package with files, imports, etc?  It seems maybe of limited use.

Reply:
Absolutely, I totally missed it.